nrf: Make the machine.uart class more compatible.

robert-hh opened the pull request:
- Add the timeout and timeout_char keyword options.
- Make uart.read() non-blocking.
- Add uart.any().
- Add ioctl MP_STREAM_POLL handling.
- Change uart.write() into non-busy waiting. uart.write() still waits until all data has been sent, but calls MICROPY_EVENT_POLL_HOOK meanwhile. uart.write() uses DMA for transfer. One option would be to add a small local buffer, such that transfers up to the size of the buffer could be done without waiting.
- As a side effect to the change of uart.write(), uart.txdone() and ioctl flush not report/wait for now correctly the end of transmission.

robert-hh committed:
nrf/modules/machine/uart: Add timeout keyword options and "any" method.

Changes in this commit:
- Add the timeout and timeout_char keyword options.
- Make uart.read() non-blocking.
- Add uart.any().
- Add ioctl MP_STREAM_POLL handling.
- Change uart.write() into non-busy waiting. uart.write() still waits until
  all data has been sent, but calls MICROPY_EVENT_POLL_HOOK while waiting.
  uart.write() uses DMA for transfer.  One option would be to add a small
  local buffer, such that transfers up to the size of the buffer could be
  done without waiting.
- As a side effect to the change of uart.write(), uart.txdone() and ioctl
  flush now report/wait correctly for the end of transmission.
- Change machine_hard_uart_buf_t in machine_hard_uart_obj_t to an instance
  of that struct, rather than a pointer to one.

dpgeorge commented on ports/nrf/modules/machine/uart.c:
```diff
@@ -323,27 +338,36 @@ STATIC MP_DEFINE_CONST_DICT(machine_hard_uart_locals_dict, machine_hard_uart_loc
     // read the data
     for (size_t i = 0; i < size; i++) {
         while (!uart_rx_any(self)) {
+            if (mp_hal_ticks_ms() > t) {  // timed out
```
This won't work: eg `t` could be 0 after a wraparound and then this will time out immediately.  Instead it needs to be `mp_hal_ticks_ms() - t0 >= timeout`

robert-hh replied:
Done.

dpgeorge commented on ports/nrf/modules/machine/uart.c:
```diff
@@ -323,27 +338,36 @@ STATIC MP_DEFINE_CONST_DICT(machine_hard_uart_locals_dict, machine_hard_uart_loc
     for (size_t i = 0; i < size; i++) {
         while (!uart_rx_any(self)) {
+            if (mp_hal_ticks_ms() > t) {  // timed out
+                if (i <= 0) {
```
`i` is unsigned so this can be `if (i == 0)`

robert-hh replied:
OK. I may have been too lazy at that moment to check the data type.

dpgeorge commented on ports/nrf/modules/machine/uart.c:
```diff
@@ -323,27 +338,36 @@ STATIC MP_DEFINE_CONST_DICT(machine_hard_uart_locals_dict, machine_hard_uart_loc
 
+    nrfx_err_t err = nrfx_uart_tx(self->p_uart, buf_in, size);
     if (err == NRFX_SUCCESS) {
+        while (machine_uart_txdone(self) == mp_const_false) {
```
I think it's clearer to just call `nrfx_uart_tx_in_progress()` directly here.  Or make a helper like `machine_uart_tx_in_progress()` that returns a `bool` and use it here and in `machine_uart_txdone()`.

robert-hh replied:
I deleted that loop, because it caused uart.write() to wait until the transmission was finished, making flush() and txdone() kind of obsolete.

robert-hh replied:
No, that loop is required. So I moved it up before the send to wait until a previous transmit is finished.

robert-hh replied:
Can't move it up, since the data is not internally buffered. So uart.write() waits until all data has been sent, and txdone() could simply return True, and flush could return w/o testing.

dpgeorge replied:
I see you made the implementation of txdone and flush still check the buffer (instead of doing nothing).  I guess that's sensible, in case things improve/change in the future with this driver and buffering.

robert-hh replied:
I removed these checks, and inserted them again. They are short, and having them in place is good for future extension. Also, with the MICROPY_EVENT_POLL_HOOK in the wait loop of uart_write a scheduled task could run run write, which may check for UART tx being finished, even if that doesn't make sense.

robert-hh replied:
PR updated & force-pushed.

dpgeorge commented on ports/nrf/modules/machine/uart.c:
```diff
@@ -355,9 +379,20 @@ STATIC mp_uint_t machine_hard_uart_write(mp_obj_t self_in, const void *buf_in, m
+        if ((flags & MP_STREAM_POLL_RD) && uart_rx_any(self) != 0) {
+            ret |= MP_STREAM_POLL_RD;
+        }
+        if ((flags & MP_STREAM_POLL_WR) && machine_uart_txdone(self) == mp_const_true) {
```
as above, probably best to make a helper `machine_uart_tx_in_progress()` (can also be used just below)

robert-hh commented on ports/nrf/modules/machine/uart.c:
```diff
@@ -99,13 +99,15 @@ typedef struct _machine_hard_uart_obj_t {
     mp_obj_base_t       base;
     const nrfx_uart_t * p_uart;      // Driver instance
     machine_hard_uart_buf_t *buf;
```
Changed it, but that caused quite many lines of code to change, mostly trivial. Testing that I found an previous bug in uart receive. It never should have worked.

robert-hh commented on ports/nrf/modules/machine/uart.c:
```diff
@@ -286,20 +292,29 @@ STATIC mp_obj_t machine_hard_uart_readchar(mp_obj_t self_in) {
 STATIC mp_obj_t machine_uart_txdone(mp_obj_t self_in) {
-    return mp_const_true;
+    machine_hard_uart_obj_t *self = self_in;
+    return nrfx_uart_tx_in_progress(self->p_uart) ? mp_const_false : mp_const_true;
```
Good hint.

dpgeorge commented on ports/nrf/modules/machine/uart.c:
```diff
@@ -321,29 +334,38 @@ STATIC const mp_rom_map_elem_t machine_hard_uart_locals_dict_table[] = {
     // read the data
     for (size_t i = 0; i < size; i++) {
         while (!uart_rx_any(self)) {
+            if ((mp_hal_ticks_ms() - t) >= self->timeout) {  // timed out
```
I don't think this is correct: eg if no chars are available after `self->timeout` then the LHS here is 0 (or maybe 1) but the condition is still false.  So it should read `(int32_t)(mp_hal_ticks_ms() - t) >= 0`.  The cast to `int32_t` is very important.

robert-hh replied:
You're right.

robert-hh committed:
nrf/nrfx_config: Use the UARTE definitions and drivers for the NRF52xx.

Suggested by @ricksorensen after testing.  These match better the hardware
of the NRF52xx.

robert-hh committed:
nrf/modules/machine: Use a dedicated function for machine.idle().

Calling MICROPY_EVENT_POLL_HOOK.  That allows Ctrl-C to break loops with
idle().

dpgeorge commented:
Thanks for updating, it looks good now.

robert-hh commented:
Thank you for merging. These changes are needed for another task to make UART timeout works consistently, but that task is stuck somewhat now at the point of adding buffered UART tx to the STM32 port. Not that it would be generally complicated, but the STM32 port carries a lot of history and legacy with it. Other ports will collect that over time too. Like ESP32, which gets more and more chaotic.
